_posts/2020-11-07-Intro-to-Regular-Expressions-and-Linux-Tools.md

```diff
@@ -0,0 +1,231 @@
+
+
+
+**Commonly Used Patterns:**
+- **.** : matches any single character
+- **[A-Z]** : matches any capitalized letter
+- **[a-z]** : matches any lower case letter
+- **^** : matches the expression at the beginning of a line
+- **$** : matches the expression at the end of a line
+- **\*** : quantifier that matches the preceding character zero or more times
+- **+** : matches the preceding character one or more time
+- **[ ]** : matches any one set of characters
+- **{n}** : matches the preceding characters *n* times
+
+Take a look at this example showing a regular expression matching an e-mail address.
+
+![regular expression example](https://ucarecdn.com/7b465367-f8ba-4033-972c-e373d278fdd2/)
+
+While regular expressions are incredibly useful, it is true they can get pretty complicated as is shown in the above example, and can thus be tricky to learn.
+
+![regular expressions meme](https://i.redd.it/fozqgb59a9p21.png)
+
+A useful tool when beginning to work with regular expressions is [Regex 101][regex101]. It is an online interactive tester that allows you to write regular expressions and test them using sample text. (In fact, I used this tool when creating the examples found later in this post!)
+
+[regex101]: https://regex101.com/
+
+There are many tools that use regular expressions as part of their functionality, including the three Linux command line tools `grep`, `sed`, and `awk`.
+
+
+# **grep**
+## The Basics
+`grep` is a command line tool that allows you to search for patterns within files. Aptly named, `grep` stands for **G**lobal **R**egular **E**xpression **P**rint. [^3]
+`grep` will find and print lines in a file that match a pattern (a specific string of characters or regular expression). One notable feature of `grep` is that it is *really fast* because it works by streaming a file's data instead of adding file data to memory all at once.
+
+So how do you use `grep`?
+
+
+## Syntax
+The basic structure of a `grep` command is shown below:
+
+```bash
+$ grep [options] "pattern" input-file
+```
+The first argument is the pattern we're searching for and the second is the input file we wish to search in.
+
+There are also a number of different options that can be used to enable other capabilities of `grep`. For instance, you can find context about the matching line using the `-Bn` option, which prints *n* lines before the matching line, and the `-An` option, which prints *n* line after the matching line.
+
+## Examples
+As a simple example, let's say we want to search for some text within an example text file: `sample.txt`.
+
+To view the contents of the `sample.txt` file, we can use the `cat` command:
+
+```bash
+$ cat sample.txt
+cat
+bat
+apple
+dog
+first
+Eat
+blue
+catnap
+Green
+```
+If we want to find all instances within `sample.txt`
+that exactly match the word "dog", we can use the following `grep` command:
```

Comment:
you used grep cat, not grep dog.
one way to make this more clear, have multiple dogs in the sample.txt so it pulls multiple lines

Reply:
Thanks for catching that! I meant to put cat. Fixed by 0d2e4bf91ae4b975327223b7231fbe1e1f08f293